From dd1bdefa80440e35a48038db1a11671b99b324d8 Mon Sep 17 00:00:00 2001
From: Andrew Boie <andrew.p.boie@intel.com>
Date: Thu, 12 Sep 2013 15:30:40 -0700
Subject: cutils: list: add list_for_each_safe

For when you want to modify the list whilst iterating over it

Issue: ADEV-21
Domain: AOSP-libcutils
Source: Internal
Upstream-candidate: Yes
https://android-review.googlesource.com/#/c/65354
Change-Id: I84432892890987c218e56883c35e52c9ff0240a3
Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
---
 include/cutils/list.h |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/include/cutils/list.h b/include/cutils/list.h
index 3881fc9..72395f4 100644
--- a/include/cutils/list.h
+++ b/include/cutils/list.h
@@ -44,6 +44,11 @@ struct listnode
 #define list_for_each_reverse(node, list) \
     for (node = (list)->prev; node != (list); node = node->prev)
 
+#define list_for_each_safe(node, next, list) \
+    for (node = (list)->next, next = node->next; \
+         node != (list); \
+         node = next, next = node->next)
+
 void list_init(struct listnode *list);
 void list_add_tail(struct listnode *list, struct listnode *item);
 void list_remove(struct listnode *item);
-- 
1.7.4.1

